src/platforms/stlinkv3/platform.c

```diff
@@ -0,0 +1,233 @@
+ *
+ * Copyright (C) 2022-2023 1BitSquared  <info@1bitsquared.com>
+ * Portions (C) 2020-2021 Stoyan Shopov <stoyan.shopov@gmail.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This file provides the platform specific functions for the STLINK-V3
+ * implementation.
+ */
+
+#include "general.h"
+#include "usb.h"
+#include "aux_serial.h"
+#include "gdb_if.h"
+
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/cm3/scb.h>
+#include <libopencm3/cm3/scs.h>
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/usart.h>
+#include <libopencm3/usb/usbd.h>
+#include <libopencm3/stm32/adc.h>
+#include <libopencm3/stm32/spi.h>
+#include <libopencm3/stm32/adc.h>
+#include <libopencm3/stm32/syscfg.h>
+
+uint16_t srst_pin;
+static uint32_t hw_version;
+
+#define SCB_CCR_IC_SHIFT               17U                       /*!< SCB CCR: Instruction cache enable bit Position */
+#define SCB_CCR_IC_MASK                (1UL << SCB_CCR_IC_SHIFT) /*!< SCB CCR: Instruction cache enable bit Mask */
+#define SCB_CCR_DC_SHIFT               16U                       /*!< SCB CCR: Cache enable bit Position */
+#define SCB_CCR_DC_MASK                (1UL << SCB_CCR_DC_SHIFT) /*!< SCB CCR: DC Mask */
+#define SCB_CCSIDR_NUMSETS_MASK        (0x7fffUL << SCB_CCSIDR_NUMSETS_SHIFT) /*!< SCB CCSIDR: NumSets Mask */
+#define SCB_CCSIDR_NUMSETS_SHIFT       13U                                    /*!< SCB CCSIDR: NumSets Position */
+#define SCB_CCSIDR_ASSOCIATIVITY_SHIFT 3U                                     /*!< SCB CCSIDR: Associativity Position */
+#define SCB_CCSIDR_ASSOCIATIVITY_MASK  (0x3ffUL << SCB_CCSIDR_ASSOCIATIVITY_SHIFT) /*!< SCB CCSIDR: Associativity Mask */
+#define CCSIDR_WAYS(x)                 (((x)&SCB_CCSIDR_ASSOCIATIVITY_MASK) >> SCB_CCSIDR_ASSOCIATIVITY_SHIFT)
+#define CCSIDR_SETS(x)                 (((x)&SCB_CCSIDR_NUMSETS_MASK) >> SCB_CCSIDR_NUMSETS_SHIFT)
+#define SCB_DCISW_SET_SHIFT            5U                               /*!< SCB DCISW: Set Position */
+#define SCB_DCISW_SET_MASK             (0x1ffUL << SCB_DCISW_SET_SHIFT) /*!< SCB DCISW: Set Mask */
+#define SCB_DCISW_WAY_SHIFT            30U                              /*!< SCB DCISW: Way Position */
+#define SCB_DCISW_WAY_MASK             (3UL << SCB_DCISW_WAY_SHIFT)     /*!< SCB DCISW: Way Mask */
+
+static void cm_dsb(void)
```

Comment:
The `dsb` and `isb` instructions are memory barriers for ensuring execution concurrency. They are actually provided as macros by CMSIS, hence the old syntax of the functions. I've now added some comments with what they do, which I hope help.

... And I hear you ask ... "But we are on a single core chip, why do we need to ensure execution concurrency?" Well due to the inherent memory ordering of all loads and stores on  Cortex-M0,  Cortex-M0+, Cortex-M3 and Cortex-M4 we don't have to have to worry about memory barrier instructions on these platform. But as this Cortex-M7 device with the Art Accelerator, all bets are off, thus the memory barriers are required.

Reply:
Ah, excellent, so to borrow from C++, this is the C-available way to implement https://en.cppreference.com/w/cpp/thread/barrier

Agreed that memory ordering is important especially to prevent the compiler re-ordering or an interrupt screwing things up here